perl/cpan/private/carton.bzl

```
@@ -0,0 +1,199 @@
+    implementation = _perl_cpan_hub_impl,
+    attrs = {
+        "hub_name": attr.string(
+            doc = "The name of the hub.",
+            mandatory = True,
+        ),
+        "modules": attr.string_list(
+            doc = "Dependencies to add to the hub.",
+            mandatory = True,
+        ),
+    },
+)
+
+_CPAN_MODULE_BUILD_FILE = """\
+load("@rules_perl//perl:perl_library.bzl", "perl_library")
+
+DEPENDENCIES = {dependencies}
+
+perl_library(
+    name = "{name}",
+    srcs = glob(
+        include = ["{output}**/*"],
+        exclude = [
+            "BUILD",
+            "WORKSPACE",
+            "*.bazel",
+            "{output}t/**/*",
+            "{output}xt/**/*",
+        ],
+    ),
+    includes = {includes},
+    deps = [
+        "@{hub_name}__" + dep
+        for dep in DEPENDENCIES
+    ],
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "{repo_name}",
+    actual = "{name}",
+    visibility = ["//visibility:public"],
+)
+"""
+
+def _perl_cpan_archive_impl(repository_ctx):
+    results = repository_ctx.download_and_extract(
+        repository_ctx.attr.urls,
+        sha256 = repository_ctx.attr.sha256,
+        stripPrefix = repository_ctx.attr.strip_prefix,
+    )
+
+    output = ""
+    includes = []
+    lib_path = repository_ctx.path("lib")
+    if lib_path.exists:
+        includes = ["lib"]
+    else:
+        # In the event a lib directory doesn't exist, the module will need to be
+        # extracted to a subdirectory.
+        repository_ctx.delete(repository_ctx.path("."))
+        output = ""
+        split = repository_ctx.attr.distribution.split("-")
+        if len(split) > 1:
+            output = split[0]
+
+        repository_ctx.download_and_extract(
+            repository_ctx.attr.urls,
+            sha256 = results.sha256,
+            stripPrefix = repository_ctx.attr.strip_prefix,
+            output = output,
+        )
+
+        includes = ["."]
+
+    repository_ctx.file("WORKSPACE.bazel", """workspace(name = "{}")""".format(
+        repository_ctx.name,
+    ))
+
+    repository_ctx.file("BUILD.bazel", _CPAN_MODULE_BUILD_FILE.format(
+        name = repository_ctx.attr.distribution,
+        hub_name = repository_ctx.attr.hub_name,
+        repo_name = "{}__{}".format(repository_ctx.attr.hub_name, repository_ctx.attr.distribution),
+        dependencies = json.encode_indent(repository_ctx.attr.dependencies, indent = " " * 4),
+        output = "{}/".format(output).lstrip("/"),
+        includes = json.encode(includes),
+    ))
+
+    return {
+        "dependencies": repository_ctx.attr.dependencies,
+        "distribution": repository_ctx.attr.distribution,
+        "hub_name": repository_ctx.attr.hub_name,
+        "name": repository_ctx.name,
+        "sha256": results.sha256,
+        "strip_prefix": repository_ctx.attr.strip_prefix,
+        "urls": repository_ctx.attr.urls,
+    }
+
+perl_cpan_archive = repository_rule(
```

Again, I'd prefer
```suggestion
cpan_archive = repository_rule(
```
or possibly
```suggestion
cpan_distribution_archive = repository_rule(
```
? Like, it's not an archive of all of CPAN ;)

---

Updated!